.github/workflows/main.yml

```diff
@@ -0,0 +1,348 @@
+name: CI
+
+on:
+  # TODO: add nightly schedule.
+  pull_request:
+    branches:
+    - master
+  push:
+    branches:
+    - master
+  release:
+    types: [published]
+
+jobs:
+  check_core:
+    name: Check Core
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python: ['3.6', '3.7']
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Setup Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python }}
+
+    - uses: actions/cache@v1
+      with:
+        path: ${{ env.pythonLocation }}
+        key: ${{ runner.os }}-pydeps-${{ matrix.python }}-${{ hashFiles('setup.py') }}-${{ hashFiles('dev-requirements.txt') }}
+        restore-keys: |
+          ${{ runner.os }}-pydeps-${{ matrix.python }}
+
+    - name: Install requirements
+      run: |
+        pip install --upgrade pip setuptools wheel
+        pip install -e .
+        pip install --upgrade -r dev-requirements.txt
+
+    - name: Debug info
+      run: |
+        pip freeze
+
+    - name: Lint
+      run: |
+        make lint
+
+    - name: Type check
+      run: |
+        make typecheck
+
+    - name: Run tests
+      run: |
+        make test-with-cov
+
+    - name: Upload coverage to Codecov
+      if: matrix.python == '3.7'
+      env:
+        CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
+      run: |
+        # Ignore codecov failures as the codecov server is not
+        # very reliable but we don't want to report a failure
+        # in the github UI just because the coverage report failed to
+        # be published.
+        # This will also fail for forked repositories since the secret token won't
+        # be available.
+        codecov -t $CODECOV_TOKEN || echo "codecov upload failed"
+
+  check_models:
+    name: Check Models
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python: ['3.6', '3.7']
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Setup Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python }}
+
+    - uses: actions/cache@v1
+      with:
+        path: ${{ env.pythonLocation }}
+        key: ${{ runner.os }}-pydeps-${{ matrix.python }}-${{ hashFiles('setup.py') }}-${{ hashFiles('dev-requirements.txt') }}
+        restore-keys: |
+          ${{ runner.os }}-pydeps-${{ matrix.python }}
+
+    - name: Install requirements
+      run: |
+        pip install --upgrade pip setuptools wheel
```

Comment:
Is it necessary to specify "setuptools" and "wheel", given that they're here and will be installed if missing by the next `pip install -e .`?:

https://github.com/allenai/allennlp/blob/0638cbb249caff0174ee0b7a6f6c8c235443e798/pyproject.toml#L22

Reply:
I did run into an issue once where some deps failed to install because wheel hadn't been installed yet. Maybe I was doing something weird at the time.

Reply:
I understand what you say. However, it should be installed now if missing because of that line I'm mentioning. See https://pip.pypa.io/en/stable/reference/pip/#pep-517-and-518-support

Reply:
Gotcha, I didn't know about that until now so thank you. It's probably a good idea to keep them up-to-date though. Do you know if `pip` will automatically do that?

Reply:
It won't, at least when I had problems with older Python versions. I'm not sure if it's necessary to keep them up-to-date but probably is necessary to set a minimum version. I had to set a minimum version for `setuptools` once, to support my features, because I saw the default in some configuration (Python 3.5?) wasn't working fine: https://github.com/pln-fing-udelar/fast-krippendorff/commit/e8ec06fa267ecaa691061e8331615f313c516566#diff-90829e76e906f1c73140c7ded7e1b268